Engine/DataFeeds/ApiDataProvider.cs

```diff
@@ -40,38 +60,114 @@ public class ApiDataProvider : IDataProvider
+                Thread.Sleep(TimeSpan.FromSeconds(3));
+            }
+            else
+            {
+                // Log URL to go accept terms
+                throw new Exception($"Must agree to terms at {_dataPrices.AgreementUrl}, before using the ApiDataProvider");
+            }
+
+            // Verify we have the balance to maintain our purchase limit, if not adjust it to meet our balance
+            _balance = organization.Credit.BalanceQCC;
+            if (_balance < _purchaseLimit)
+            {
+                Log.Trace("ApiDataProvider(): Purchase limit is greater than balance." +
+                    $" Setting purchase limit to balance : {_balance}");
+                _purchaseLimit = _balance;
+            }
         }
 
         /// <summary>
         /// Retrieves data to be used in an algorithm.
         /// If file does not exist, an attempt is made to download them from the api
         /// </summary>
-        /// <param name="key">A string representing where the data is stored</param>
+        /// <param name="filePath">File path representing where the data requested</param>
         /// <returns>A <see cref="Stream"/> of the data requested</returns>
-        public Stream Fetch(string key)
+        public Stream Fetch(string filePath)
         {
-            Symbol symbol;
-            DateTime date;
-            Resolution resolution;
-
-            // Fetch the details of this data request
-            if (LeanData.TryParsePath(key, out symbol, out date, out resolution))
+            // If we don't already have this file or its out of date, download it
+            if (NeedToDownload(filePath))
             {
-                if (!File.Exists(key) || IsOutOfDate(resolution, key))
+                // TODO Factor files and map files do not have a price, so they come back as -1. This works for now but do we want to try and handle that case uniquely?
+                // Verify we have enough credit to handle this
+                var price = _dataPrices.GetPrice(filePath);
+                if (price > 0)
                 {
-                    return DownloadData(key, symbol, date, resolution);
+                    if (_purchaseLimit < price)
+                    {
+                        throw new Exception($"Cost {price} for {filePath} data exceeds remaining purchase limit: {_purchaseLimit}");
+                    }
+
+                    // Update our purchase limit and balance.
+                    _purchaseLimit -= price;
+                    _balance -= price;
                 }
 
-                // Use the file already on the disk
-                return new FileStream(key, FileMode.Open, FileAccess.Read, FileShare.Read);
+                return DownloadData(filePath);
+            }
+
+            // Use the file already on the disk
+            if (File.Exists(filePath))
+            {
+                return new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
```

Comment:
Maybe it could inherit from `DefaultDataProvider` and call the base implementation here which skips checking if the file exists on disk, since it's already done by `FileStream`